### PR TITLE
MF-1510 - Add Event Sourcing To Users Service

### LINKED ISSUE #1510: Event sourcing for groups
**FEATURE REQUEST**

2. Describe the feature you are requesting, as well as the possible use case(s) for it.
Add event sourcing to groups.  

3. Indicate the importance of this feature to you (must-have, should-have, nice-to-have).
should-have

This would work in the same way as event sourcing for things, using Redis Streams for all of the appropriate operations made on groups.

### PR DESCRIPTION
**Pull request by rodneyosodo**
Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

### What does this do?
Emit events from `users` service

### Which issue(s) does this PR fix/relate to?
Resolves #1510

### List any changes that modify/break current functionality
Add event sourcing on `users` services. I have added for both `clients`, `groups` and `policies` sub-services

### Have you included tests for your changes?
No

### Did you document any new/modified functionality?
No

### Notes
image

### CODE REVIEW

**Review by SammyOina — APPROVED** [review]
LGTM

File: cmd/users/main.go [hunk 1]
```diff
@@ -32,10 +34,12 @@ import (
 	capi "github.com/mainflux/mainflux/users/clients/api"
 	"github.com/mainflux/mainflux/users/clients/emailer"
 	cpostgres "github.com/mainflux/mainflux/users/clients/postgres"
+	redisucache "github.com/mainflux/mainflux/users/clients/redis"
```
Comment (dborovcanin) [inline cmd/users/main.go]: You can replace it with `ucache`, and do the same for `gcache` for groups and `pcache` for policies. (I will create a ticket to address this in all the `main.go` files)

File: things/clients/redis/streams.go [hunk 2]
```diff
@@ -59,89 +59,89 @@ func (es eventStore) CreateThings(ctx context.Context, token string, thing ...mf
 }
 
 func (es eventStore) UpdateClient(ctx context.Context, token string, thing mfclients.Client) (mfclients.Client, error) {
-	cli, err := es.svc.UpdateClient(ctx, token, thing)
+	thing, err := es.svc.UpdateClient(ctx, token, thing)
```
Comment (dborovcanin) [inline things/clients/redis/streams.go]: No need for this rename, we can keep the `clients` terminology.

File: things/groups/redis/streams.go [hunk 3]
```diff
@@ -32,36 +32,36 @@ func NewEventStoreMiddleware(svc groups.Service, client *redis.Client) groups.Se
 	}
 }
 
-func (es eventStore) CreateGroups(ctx context.Context, token string, groups ...mfgroups.Group) ([]mfgroups.Group, error) {
-	gs, err := es.svc.CreateGroups(ctx, token, groups...)
+func (es eventStore) CreateGroups(ctx context.Context, token string, grps ...mfgroups.Group) ([]mfgroups.Group, error) {
```
Comment (dborovcanin) [inline things/groups/redis/streams.go]: With the same analogy, we could replace `token` -> `tkn`. For the time being, let's not rename (at least in this PR).

**Review by dborovcanin — CHANGES_REQUESTED** [review]

**Review by dborovcanin — CHANGES_REQUESTED** [review]
Please fix User events sourcing, CI is failing.

File: users/policies/redis/events.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+		val["action"] = ae.Action
+	}
+	return val, nil
+}
+
+type listPoliciesEvent struct {
+	policies.Page
+}
+
+func (ae listPoliciesEvent) Encode() (map[string]interface{}, error) {
```
Comment (drasko) [inline users/policies/redis/events.go]: This should probably be `le`, not `ae`

File: users/groups/redis/events.go [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+	}
+
+	return val, nil
+}
+
+type listGroupEvent struct {
+	mfgroups.GroupsPage
+}
+
+func (lce listGroupEvent) Encode() (map[string]interface{}, error) {
```
Comment (drasko) [inline users/groups/redis/events.go]: `le` would probably be better, or `lge`, I do not see where `c` comes from.

**Review by drasko — CHANGES_REQUESTED** [review]

**Review by drasko — DISMISSED** [review]
LGTM

**Pull request merged**